Use completion rates to decide injection/production startpoints.

atgeirr opened the pull request:
This is instead of using the declared well type (injector/producer), since that will add zero-rate perforation cells to the start sets and also cross-flowing perforation cells.

With this, crossflowing wells' solutions will be output twice: both for injecting and producing diagnostics.

Has been tested for Norne step 27: `computeLocalSolutions case=NORNE_ATW2013 step=27` (and step 206) with successful runs (as opposed to stopping due to mismatch between inflow fluxes and start points).

bska commented:
Looks okay from an cursory glance.  I think I'd probably like to keep the `activeCompletions()` helper though--mainly to avoid deeply nested control structures.  That would further enable putting criteria on **both** the well and its completions as outlined below.  I'm not suggesting that we add such criteria now, but I'd nevertheless like to have the option at a future date.

---

 ```C++
 template <class WellData, class PickCompl>
 std::vector<int>
 activeCompletions(const ::Opm::ECLGraph& G,
                   const WellData&        well,
                   PickCompl&&            pickCompl)
 {
    auto completion_cells = std::vector<int>{};

    completion_cells.reserve(well.completions.size());

    for (const auto& completion : well.completions) {
        if (! pickCompl(completion)) { continue; }

        const auto cell_index =
            G.activeCell(completion.ijk, completion.gridName);

        if (cell_index >= 0) {
            completion_cells.push_back(cell_index);
        }
    }

    return completion_cells;
}

template <class SelectWell, class SelectCompl>
StartSets
getStartPointsFromWells(const example::Setup& setup,
                         SelectWell&&          pickWell,
                         SelectCompl&&         pickCompl)
{
    auto start = StartSets{};

    for (const auto& well : setup.well_fluxes) {
        if (! pickWell(well)) { continue; }

        auto completions =
            activeCompletions(setup.graph, well,
                              std::forward<SelectCompl>(pickCompl));

        if (completions.empty()) { continue; }

        start.emplace_back(Opm::FlowDiagnostics::CellSetID(well.name),
                           std::move(completions));
    }

    return start;
}
```

atgeirr commented:
I have modified the code to split the control between wells and completions as requested, as well as a few other minor changes.

I see in the diff that I seemingly made some cosmetic changes, they were not really intended as such, but happened because to make the "new" activeCompletions() method I moved code from the get...() method rather than dig up the old one...

bska commented:
> I have modified the code to split the control between wells and completions as requested, as well as a few other minor changes.

I like it.  Thanks a lot.  I'll merge into master.
